integration-tests/tests/src/serialization.ts

```diff
@@ -0,0 +1,89 @@
+import { expect } from "chai";
+
+import { IPerson, PersonSchema } from "./schemas/person-and-dogs";
+import * as expectedCircularStructure from "./structures/circular-result.json"
+
+describe("Serialization", () => {
+    it("JsonSerializationReplacer is exposed on the Realm constructor", () => {
+        expect(typeof Realm.JsonSerializationReplacer).equal('function');
+        expect(Realm.JsonSerializationReplacer.length).equal(2);
+    })
+
+    it("Realm.Object has 'toJSON' implemented", () => {
+        const realm = new Realm({ schema: [PersonSchema] });
+
+        let john!: IPerson & Realm.Object;
+
+        realm.write(() => {
+            john = realm.create<IPerson>("Person", {
+                name: "John Doe",
+                age: 42
+            });
+
+            john.friends.push(john)
+        });
+
+        expect(typeof john.toJSON).equals('function');
+        expect(john.toJSON().objectSchema).equals(undefined);
+        expect(john.toJSON().friends).instanceOf(Array);
+    });
+
+    it("Realm.Results (Collection) has 'toJSON' implemented", () => {
+        const realm = new Realm({ schema: [PersonSchema] });
+
+        let john!: IPerson & Realm.Object;
+
+        realm.write(() => {
+            john = realm.create<IPerson>("Person", {
+                name: "John Doe",
+                age: 42
+            });
+        });
+
+        const persons = realm.objects<IPerson>("Person");
+
+        expect(typeof persons.toJSON).equals('function');
+        expect(persons.toJSON()).instanceOf(Array);
+
+        const first = persons.toJSON()[0];
+        expect(first.objectSchema).equals(undefined);
+        expect(first.friends).instanceOf(Array);
+    });
```

Comment:
It would be great with a text which explicitly tests the behaviour `toJSON` with circular references.
Specifically I would expect `john` to be the same object (by reference) if he is befriended with himself (or via other person).

Reply:
included in the updated tests.